src/helpers.coffee

```diff
@@ -111,12 +111,34 @@ buildLocationData = (first, last) ->
 # This returns a function which takes an object as a parameter, and if that
 # object is an AST node, updates that object's locationData.
 # The object is returned either way.
-exports.addLocationDataFn = (first, last) ->
+exports.addDataToNode = (parserState, first, last) ->
   (obj) ->
-    if ((typeof obj) is 'object') and (!!obj['updateLocationDataIfMissing'])
+    # Add location data
+    if obj?.updateLocationDataIfMissing? and first?
       obj.updateLocationDataIfMissing buildLocationData(first, last)
 
-    return obj
+    # Add comments data
+    unless parserState.tokenComments
+      parserState.tokenComments = {}
+      for token in parserState.parser.tokens when token.comments
+        tokenHash = "#{token[2].first_line}x#{token[2].first_column}-#{token[2].last_line}x#{token[2].last_column}"
+        unless parserState.tokenComments[tokenHash]?
```

Comment:
I try to put the most likely condition first. In this case, it’s much more likely that we haven’t yet encountered this hash before, so that’s the primary flow and the `else` is the exception.